extensions/attributegen/plugin_test.cc

```diff
@@ -0,0 +1,489 @@
+      EXPECT_EQ(absl::SimpleAtoi(resp_code, &response_code), true);
+    }
+
+    ON_CALL(encoder_callbacks_.stream_info_, responseCode())
+        .WillByDefault(Invoke([response_code]() { return response_code; }));
+
+    EXPECT_EQ(Http::FilterHeadersStatus::Continue,
+              filter_->decodeHeaders(request_headers, true));
+
+    Buffer::OwnedImpl data(bdata);
+    EXPECT_EQ(Http::FilterDataStatus::Continue,
+              filter_->decodeData(data, true));
+
+    EXPECT_EQ(Http::FilterHeadersStatus::Continue,
+              filter_->encodeHeaders(response_headers, true));
+
+    filter_->log(&request_headers, nullptr, nullptr, request_stream_info_);
+    return fs;
+  }
+
+  // Many of the following are not used yet, but are useful
+  Stats::IsolatedStoreImpl stats_store_;
+  Stats::ScopeSharedPtr scope_;
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  NiceMock<Runtime::MockRandomGenerator> random_;
+  NiceMock<Upstream::MockClusterManager> cluster_manager_;
+  NiceMock<Init::MockManager> init_manager_;
+  WasmHandleSharedPtr wasm_;
+  PluginSharedPtr plugin_;
+  std::unique_ptr<TestFilter> filter_;
+  NiceMock<Envoy::Ssl::MockConnectionInfo> ssl_;
+  NiceMock<Envoy::Network::MockConnection> connection_;
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> decoder_callbacks_;
+  NiceMock<Http::MockStreamEncoderFilterCallbacks> encoder_callbacks_;
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> request_stream_info_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  envoy::config::core::v3::Metadata listener_metadata_;
+  TestRoot* root_context_ = nullptr;
+  Config::DataSource::RemoteAsyncDataProviderPtr remote_data_provider_;
+};
+
+}  // namespace Wasm
+}  // namespace HttpFilters
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace AttributeGen {
+using HttpFilters::Wasm::ConfigParams;
+using HttpFilters::Wasm::TestParams;
+using HttpFilters::Wasm::WasmHttpFilterTest;
+
+std::vector<TestParams> generateTestParams() {
+  return std::vector<TestParams>{
+      {.runtime = "null", .testdata_dir = "/extensions/attributegen/testdata"},
+      // {.runtime = "v8", .testdata_dir =
+      // "/extensions/attributegen/testdata"},
+  };
+}
+
+class AttributeGenFilterTest : public WasmHttpFilterTest {
+ public:
+  void verifyRequest(Http::TestRequestHeaderMapImpl& request_headers,
+                     Http::TestResponseHeaderMapImpl& response_headers,
+                     const std::string& attribute, bool found, bool error,
+                     const std::string& value = "") {
+    auto fs = makeTestRequest(request_headers, response_headers);
+
+    ASSERT_EQ(fs->hasData<WasmState>(attribute + "__error"), error);
+    ASSERT_EQ(fs->hasData<WasmState>(attribute), found)
+        << absl::StrCat(attribute, "=?", value);
+    if (found) {
+      ASSERT_EQ(fs->getDataReadOnly<WasmState>(attribute).value(), value)
```

Comment:
Since 1.14.1 PR merged, you want to update to `"wasm." + attribute` here I think.